src/os/LMDBStore.cc

```diff
@@ -0,0 +1,482 @@
+  cct->get_perfcounters_collection()->add(logger);
+  return 0;
+}
+
+int LMDBStore::_test_init(const string& dir)
+{
+  MDB_env *_env;
+  MDB_txn *_txn;
+  MDB_dbi _dbi;
+  int rc;
+  string k = "key_test";
+  string v = "value_test";
+  MDB_val key,value;
+  key.mv_size = k.size();
+  key.mv_data = (void *)(k.data());
+  value.mv_size = v.size();
+  value.mv_data = (void *)(v.data());
+  rc = mdb_env_create(&_env);
+  if (rc != 0) {
+    derr << __FILE__ << ":" << __LINE__ << " " << mdb_strerror(rc) << dendl;
+    return -1;
+  }
+  rc = mdb_env_open(_env, dir.c_str(), MDB_FIXEDMAP, 0644);
+  if (rc != 0) {
+    derr << __FILE__ << ":" << __LINE__ << " " << mdb_strerror(rc) << dendl;
+    mdb_env_close(_env);
+    return -EIO;
+  }
+  rc = mdb_txn_begin(_env, NULL, 0, &_txn);
+  if (rc != 0) {
+    derr << __FILE__ << ":" << __LINE__ << " " << mdb_strerror(rc) << dendl;
+    mdb_txn_abort(_txn);
+    mdb_env_close(_env);
+    return -EIO;
+  }
+  rc = mdb_dbi_open(_txn, NULL, 0, &_dbi);
+  if (rc != 0) {
+    derr << __FILE__ << ":" << __LINE__ << " " << mdb_strerror(rc) << dendl;
+    mdb_txn_abort(_txn);
+    mdb_env_close(_env);
+    return -EIO;
+  }
+  rc = mdb_put(_txn, _dbi, &key, &value, MDB_NOOVERWRITE);
+  if (rc != 0) {
+    derr << __FILE__ << ":" << __LINE__ << " " << mdb_strerror(rc) << dendl;
+    mdb_txn_abort(_txn);
+    mdb_dbi_close(_env, _dbi);
+    mdb_env_close(_env);
+    return -EIO;
+  }
+  rc = mdb_del(_txn, _dbi, &key, NULL);
+  if (rc != 0) {
+    derr << __FILE__ << ":" << __LINE__ << " " << mdb_strerror(rc) << dendl;
+    mdb_txn_abort(_txn);
+    mdb_dbi_close(_env, _dbi);
+    mdb_env_close(_env);
+    return -EIO;
+  }
+  rc = mdb_txn_commit(_txn);
+  if (rc != 0) {
+    derr << __FILE__ << ":" << __LINE__ << " " << mdb_strerror(rc) << dendl;
+    mdb_txn_abort(_txn);
+    mdb_dbi_close(_env, _dbi);
+    mdb_env_close(_env);
+    return -EIO;
+   }
+  mdb_dbi_close(_env, _dbi);
+  mdb_env_close(_env);
+  return (rc == 0) ? 0 : -EIO;
+} 
+
+LMDBStore::LMDBStore(CephContext *c, const string &path) :
+  cct(c),
+  logger(NULL),
+  path(path),
+  dbi(0),
+  options()
+{
+  MDB_env *_env;
+  int rc = mdb_env_create(&_env);
+  if (rc != 0) {
+    derr << __FILE__ << ":" << __LINE__ << dendl;
+  }
+  env.reset(_env, mdb_env_close);
+}
+
+LMDBStore::~LMDBStore()
+{
+  close();
+  mdb_dbi_close(env.get(), dbi);
+  if (env.get()) {
+    env.reset();
+  }
+  delete logger;
+}
+
+void LMDBStore::close()
+{
+  if (logger)
+    cct->get_perfcounters_collection()->remove(logger);
+}
+
+int LMDBStore::submit_transaction(KeyValueDB::Transaction t)
+{ 
+  int rc;
+  LMDBTransactionImpl * _t =
+    static_cast<LMDBTransactionImpl *>(t.get());
+  rc = mdb_txn_commit(_t->txn);
+  logger->inc(l_lmdb_txns);
+  return (rc == 0) ? 0 : -1;
+}
+
+int LMDBStore::submit_transaction_sync(KeyValueDB::Transaction t)
+{
+  int rc;
+  LMDBTransactionImpl * _t =
+    static_cast<LMDBTransactionImpl *>(t.get());
+  rc = mdb_txn_commit(_t->txn);
```

Comment:
Hi xinxin,

From mdb_txn_commit(http://symas.com/mdb/doc/group__mdb.html#ga846fbd6f46105617ac9f4d76476f6597), it seemed that lmdb only write userspace memory to system buffer(OS page cache). Although this could prevent data loss if application crash.

We still need to call mdb_env_sync(http://symas.com/mdb/doc/group__mdb.html#ga85e61f05aa68b520cc6c3b981dba5037) to ensure system buffer flush to disk. :-)